tests/merkle_consistency/consistency.sol

```diff
@@ -0,0 +1,223 @@
+        // Hash all leafs
+        for (uint256 i = 0; i < data.length; i++) {
+            data[i] = sha256(abi.encodePacked(uint256(0), data[i]));
+        }
+
+        uint256 step = 2;
+        while (step / 2 < data.length) {
+            for (uint256 i = 0; i < data.length - step / 2; i += step) {
+                data[i] = sha256(
+                    abi.encodePacked(uint256(1), data[i], data[i + step / 2])
+                );
+            }
+            step *= 2;
+        }
+
+        return data[0];
+    }
+
+    // Returns the merkle proof of node indicated by _index in data
+    function path(bytes32[] memory data, uint256 _index)
+        public
+        pure
+        returns (bytes32[] memory, bool[] memory)
+    {
+        uint256 index = _index;
+        for (uint256 i = 0; i < data.length; i++) {
+            data[i] = sha256(abi.encodePacked(uint256(0), data[i]));
+        }
+
+        uint256 proofIndex;
+        uint256 proofLength = log2Ceiling(data.length);
+        bytes32[] memory proof = new bytes32[](proofLength);
+        bool[] memory siblings = new bool[](proofLength);
+
+        uint256 step = 2;
+        while (step / 2 < data.length) {
+            for (uint256 i = 0; i < data.length - step / 2; i += step) {
+                if (i == index) {
+                    proof[proofIndex] = data[i + step / 2];
+                    siblings[proofIndex] = true;
+                    proofIndex++;
+                    index = i;
+                } else if (i + step / 2 == index) {
+                    proof[proofIndex] = data[i];
+                    siblings[proofIndex] = false;
+                    proofIndex++;
+                    index = i;
+                }
+                data[i] = sha256(
+                    abi.encodePacked(uint256(1), data[i], data[i + step / 2])
+                );
+            }
+            step *= 2;
+        }
+
+        // If the proof was smaller than log2(data.length), remove zero cells
+        // proofIndex points the first zero record, or the end of the proof
+        if (proofIndex != proof.length) {
+            proof = arrays.subArrayBytes32(proof, 0, proofIndex);
+            siblings = arrays.subArrayBool(siblings, 0, proofIndex);
+        }
+
+        return (proof, siblings);
+    }
+
+    // Returns the merkle tree root as calculated from a proof
+    function rootFromPath(
+        bytes32 nodeData,
+        bytes32[] memory proof,
+        bool[] memory siblings
+    ) public pure returns (bytes32) {
+        bytes32 h = sha256(abi.encodePacked(uint256(0), nodeData));
+        bytes32 left;
+        bytes32 right;
+
+        for (uint256 i = 0; i < siblings.length; i++) {
+            if (siblings[i] == true) {
+                left = h;
+                right = proof[i];
+            } else {
+                left = proof[i];
+                right = h;
+            }
+            h = sha256(abi.encodePacked(uint256(1), left, right));
+        }
+        return h;
+    }
```

Comment:
you don't really need left/right bits as they can be inferred from n if you pass it in